library/ssl_srv.c

```diff
@@ -1870,10 +1870,19 @@ static int ssl_parse_client_hello( mbedtls_ssl_context *ssl )
         return( MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
     }
 
+#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+    /*
+     * Server certification selection (after processing TLS extensions)
+     */
+    if( ssl->conf->f_cert_cb && ( ret = ssl->conf->f_cert_cb( ssl ) ) != 0 )
+        return( ret );
```

Comment:
I think it would be good to have a `MBEDTLS_SSL_DEBUG_RET()` before we return.

Note: if you agree with my suggestion, this is where we'd set `sni_name = NULL` too.

Reply:
:rocket: made both changes.  It makes sense to invalidate `ssl->handshake->sni_name` here to avoid it pointing to invalid data when after a new client message is received.

Reply:
Would we not want to clear sni_name and sni_name_len on failure as well?

Reply:
What about the 17 calls to `return` above that point in the code and which might be called after sni_name and sni_name_len are set?  If continuing the handshake is permitted after an error is returned -- is this allowed? -- then all calls to return should call a (cold) "reset" function which resets internal values before passing through the error.

Reply:
Attempting to continue the handshake after an error is not permitted and the documentation [explicitly says so](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/ssl.h#L4284). So, I think based on that, it's OK to clear those only on the success path, where the handshake will continue, but not on the error path, because well-behaved applications will call `ssl_free()` or `ssl_session_reset()` which should do that (and misbehaved applications have way bigger issues).

Which reminds me I had completely forgotten to check if the new fields in in `struct mbedtls_ssl_handshake_params` are correctly initialized/freed/reset: fortunately they are, thanks to the big `memset()` call at the beginning of `ssl_handshake_params_init()` (called indirectly by `ssl_session_reset()`) and the big `mbedtls_platform_zeroize()` call at the end of `mbedtls_ssl_handshake_free()` (called by `ssl_free()`). Phew. Thanks Paul for bringing up the error case.

Reply:
:rocket: